include/layers/Tensor.hpp

```diff
@@ -0,0 +1,166 @@
+#pragma once
+
+#include <cstdint>
+#include <iostream>
+#include <stdexcept>
+#include <type_traits>
+#include <vector>
+
+#include "layers/Shape.hpp"
+
+enum class Type { kUnknown, kInt, kDouble };
+
+template <typename T>
+std::vector<uint8_t>* to_byte(std::vector<T>& v) {
+  return reinterpret_cast<std::vector<uint8_t>*>(&v);
+}
+
+template <typename T>
+const std::vector<uint8_t>* to_byte(const std::vector<T>& v) {
+  return reinterpret_cast<const std::vector<uint8_t>*>(&v);
+}
+
+template <typename T>
+Type GetTypeEnum() {
+  if constexpr (std::is_same<T, int>::value) {
+    return Type::kInt;
+  } else if constexpr (std::is_same<T, double>::value) {
+    return Type::kDouble;
+  } else {
+    return Type::kUnknown;
+  }
+}
+
+class Tensor {
+ private:
+  Shape shape_;
+  std::vector<uint8_t> values_;
+  Type type_;
+
+  std::vector<uint8_t> SetRightTypeValues() {
+    if (type_ == Type::kInt) {
+      return std::vector<uint8_t>(shape_.count() * sizeof(int), 0);
+    }
+    if (type_ == Type::kDouble) {
+      return std::vector<uint8_t>(shape_.count() * sizeof(double), 0);
+    }
+    return std::vector<uint8_t>();
+  }
+
+  template <typename T>
+  std::vector<T>* as();
+
+  template <typename T>
+  const std::vector<T>* as() const;
+
+ public:
+  Tensor(const std::vector<uint8_t>& a, const Shape& s, Type type) {
+    type_ = type;
+    shape_ = s;
+    values_ = SetRightTypeValues();
+
+    if (a.size() != values_.size() || type == Type::kUnknown) {
+      throw std::invalid_argument("Wrong_Length");
+    }
+    values_ = a;
+  }
+
+  Tensor(const std::vector<size_t>& dims, Type type) : shape_(dims) {
+    type_ = type;
+    values_ = SetRightTypeValues();
+  }
+
+  Tensor(const Shape& sh, Type type) : shape_(sh) {
+    type_ = type;
+    values_ = SetRightTypeValues();
+  }
+
+  Tensor(const Tensor& t) = default;
+
+  Tensor(Tensor&& t) = default;
+
+  Tensor& operator=(Tensor&& t) = default;
+
+  std::vector<uint8_t> get_values() const { return values_; }
+  Shape get_shape() const { return shape_; }
+  Type get_type() const noexcept { return type_; }
+
+  template <typename T>
+  typename std::vector<T>::const_iterator begin() const {
+    return this->as<T>().begin();
+  }
+
+  template <typename T>
+  typename std::vector<T>::const_iterator end() const {
+    return this->as<T>().end();
+  }
+
+  template <typename T>
+  T& get(const std::vector<size_t>& coords);  // write
```

Comment:
if we use write/read then I think using set/get notation will be more infromative

Reply:
Fixed